File: my-code/345/main.go
package main

// @Time     : 2021/5/31 10:22
// @Author   : Ranshi
// @File     : 345. 反转字符串中的元音字母.go

func reverseVowels(s string) string {
	oriCheck := map[byte]bool{
		'a': true, 'o': true, 'e': true, 'i': true, 'u': true, 'A': true, 'O': true, 'E': true, 'I': true, 'U': true,
	}
	b := []byte(s)
	for l, r := 0, len(b)-1; l < r; {
		for _, ok := oriCheck[b[r]]; l < r && !ok; ok = oriCheck[b[r]] {
			r--
		}
		for _, ok := oriCheck[b[l]]; l < r && !ok; ok = oriCheck[b[l]] {
			l++
		}
		b[l], b[r] = b[r], b[l]
		l++
		r--
	}
	return string(b)
}

File: my-code/345/main.py
# -*- coding: UTF-8 -*-
# @Time     : 2021/08/19 07:19
# @Author   : Ranshi
# @File     : 123.py


class Solution:
    def reverseVowels(self, s: str) -> str:
        vowel = {"a", "o", "e", "i", "u", "A", "O", "E", "I", "U"}
        chs = list(s)
        lo, hi = 0, len(chs) - 1
        while lo < hi:
            while chs[lo] not in vowel and lo < hi:
                lo += 1
            while chs[hi] not in vowel and lo < hi:
                hi -= 1
            chs[lo], chs[hi] = chs[hi], chs[lo]
            lo, hi = lo + 1, hi - 1
        return "".join(chs)


if __name__ == "__main__":
    print(Solution().reverseVowels("ai"))
